libs/geo/src/main/java/org/elasticsearch/geometry/utils/SpatialEnvelopeVisitor.java

```diff
@@ -0,0 +1,302 @@
+ *         Rectangle envelope = bbox.get();
+ *         // Do stuff with the envelope
+ *     }
+ * </code>
+ * It is also possible to create the inner <code>PointVisitor</code> separately, as well as use the visitor for multiple geometries.
+ * <code>
+ *     PointVisitor pointVisitor = new CartesianPointVisitor();
+ *     SpatialEnvelopeVisitor visitor = new SpatialEnvelopeVisitor(pointVisitor);
+ *     for (Geometry geometry : geometries) {
+ *         geometry.visit(visitor);
+ *     }
+ *     if (visitor.isValid()) {
+ *         Rectangle envelope = visitor.getResult();
+ *         // Do stuff with the envelope
+ *     }
+ * </code>
+ * Code that wishes to modify the behaviour of the visitor can implement the <code>PointVisitor</code> interface,
+ * or extend the existing implementations.
+ */
+public class SpatialEnvelopeVisitor implements GeometryVisitor<Boolean, RuntimeException> {
+
+    private final PointVisitor pointVisitor;
+
+    public SpatialEnvelopeVisitor(PointVisitor pointVisitor) {
+        this.pointVisitor = pointVisitor;
+    }
+
+    /**
+     * Determine the BBOX without considering the CRS or wrapping of the longitude.
+     * Note that incoming BBOX's that do cross the dateline (minx>maxx) will be treated as invalid.
+     */
+    public static Optional<Rectangle> visit(Geometry geometry) {
+        var visitor = new SpatialEnvelopeVisitor(new CartesianPointVisitor());
+        if (geometry.visit(visitor)) {
+            return Optional.of(visitor.getResult());
+        }
+        return Optional.empty();
+    }
+
+    /**
+     * Determine the BBOX assuming the CRS is geographic (eg WGS84) and optionally wrapping the longitude around the dateline.
+     */
+    public static Optional<Rectangle> visit(Geometry geometry, boolean wrapLongitude) {
+        var visitor = new SpatialEnvelopeVisitor(new GeoPointVisitor(wrapLongitude));
+        if (geometry.visit(visitor)) {
+            return Optional.of(visitor.getResult());
+        }
+        return Optional.empty();
+    }
+
+    public Rectangle getResult() {
+        return pointVisitor.getResult();
+    }
+
+    /**
+     * Visitor for visiting points and rectangles. This is where the actual envelope calculation happens.
+     * There are two implementations, one for cartesian coordinates and one for geographic coordinates.
+     * The latter can optionally wrap the longitude around the dateline.
+     */
+    public interface PointVisitor {
+        void visitPoint(double x, double y);
+
+        void visitRectangle(double minX, double maxX, double maxY, double minY);
+
+        boolean isValid();
+
+        Rectangle getResult();
+    }
+
+    /**
+     * The cartesian point visitor determines the envelope by the minimum and maximum x/y coordinates.
+     * It also disallows invalid rectangles where minX > maxX.
+     */
+    public static class CartesianPointVisitor implements PointVisitor {
+        private double minX = Double.POSITIVE_INFINITY;
```

Comment:
These should be either `protected` or be exposed via getters so they can be used by its sub classes, e.g., during intermediate serialization.

Reply:
Got this from the changes in ST_EXTENT PR.